openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,323 @@
+	podWorkloadAnnotationEffect             = "effect"
+	workloadEffectPreferredDuringScheduling = "PreferredDuringScheduling"
+	managedCapacityLabel                    = "management.workload.openshift.io/cores"
+	containerResourcesAnnotationPrefix      = "io.openshift.workload"
+	containerCPUSharesAnnotation            = "io.openshift.workload.management.cpushares"
+	containerCPUSetAnnotation               = "io.openshift.workload.management.cpuset"
+	namespaceAllowedAnnotation              = "workload.openshift.io/allowed"
+	workloadTypeManagement                  = "management"
+)
+
+var _ = initializer.WantsExternalKubeInformerFactory(&managementCPUsOverride{})
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+// managementCPUsOverride presents admission plugin that will replace pod containers CPUs requests to a new management
+// resource
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister   corev1listers.NamespaceLister
+	nodeLister corev1listers.NodeLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace %q: %v", PluginName, attr.GetNamespace(), err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	if !isPodManaged(pod.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
```

Comment:
why not stripping workload annotations here?

Reply:
done, moved the stripping functionality out of the `if`